python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,292 @@
+
+  return block_max;
+}
+
+template <typename T>
+__device__ T BlockMin(T const* data, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  auto local_min = std::numeric_limits<T>::max();
+  __shared__ T block_min;
+  if (block.thread_rank() == 0) { block_min = local_min; }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    local_min = min(local_min, data[idx]);
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{block_min};
+  ref.fetch_min(local_min, cuda::std::memory_order_relaxed);
+
+  block.sync();
+
+  return block_min;
+}
+
+template <typename T>
+__device__ int64_t BlockIdxMax(T const* data, int64_t* index, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_max;
+  __shared__ int64_t block_idx_max;
+
+  auto local_max = []() {
+    if constexpr (std::is_floating_point_v<T>) { return -std::numeric_limits<T>::max(); }
+    return std::numeric_limits<T>::min();
+  }();
+  auto local_idx_max = std::numeric_limits<int64_t>::max();
+
+  if (block.thread_rank() == 0) {
+    block_max     = local_max;
+    block_idx_max = local_idx_max;
+  }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto const current_data = data[idx];
+    if (current_data > local_max) {
+      local_max     = current_data;
+      local_idx_max = index[idx];
+    }
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{block_max};
+  ref.fetch_max(local_max, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  if (local_max == block_max) {
+    cuda::atomic_ref<int64_t, cuda::thread_scope_block> ref_idx{block_idx_max};
+    ref_idx.fetch_min(local_idx_max, cuda::std::memory_order_relaxed);
+  }
+  block.sync();
+
+  return block_idx_max;
+}
+
+template <typename T>
+__device__ int64_t BlockIdxMin(T const* data, int64_t* index, int64_t size)
+{
+  auto block = cooperative_groups::this_thread_block();
+
+  __shared__ T block_min;
+  __shared__ int64_t block_idx_min;
+
+  auto local_min     = std::numeric_limits<T>::max();
+  auto local_idx_min = std::numeric_limits<int64_t>::max();
+
+  if (block.thread_rank() == 0) {
+    block_min     = local_min;
+    block_idx_min = local_idx_min;
+  }
+  block.sync();
+
+#pragma unroll
+  for (int64_t idx = block.thread_rank(); idx < size; idx += block.size()) {
+    auto const current_data = data[idx];
+    if (current_data < local_min) {
+      local_min     = current_data;
+      local_idx_min = index[idx];
+    }
+  }
+
+  cuda::atomic_ref<T, cuda::thread_scope_block> ref{block_min};
+  ref.fetch_min(local_min, cuda::std::memory_order_relaxed);
+  block.sync();
+
+  if (local_min == block_min) {
+    cuda::atomic_ref<int64_t, cuda::thread_scope_block> ref_idx{block_idx_min};
+    ref_idx.fetch_min(local_idx_min, cuda::std::memory_order_relaxed);
+  }
+  block.sync();
+
+  return block_idx_min;
+}
```

Comment:
Not to be acted on, since I think the long-term plan is to switch to CUB, but this copy-paste is hiding a monoid abstraction that could be used to reduce duplication.